syntax = "proto3";

// Data required for resuming a training process
message CheckpointParams {
    int32 version = 25;                                 // Rolling version

    ModelParams model = 1;

    float checkpoint_frequency = 8;
    string output_dir = 9;
    string output_model_prefix = 22;

    bool data_aug_retrain_on_original = 23;        // Retrain the model with only the non augmented data in a second run
    int32 current_stage = 24;                      // Current training progress: 0 standard, 1 retraining on non aug.

    // params for early stopping
    float early_stopping_at_acc = 26;                   // Stop training if validation reaches this accuracy threshold
    float early_stopping_frequency = 15;                // How often to check on validation data
    int32 early_stopping_nbest = 16;                    // How many models must be worse to stop
    string early_stopping_best_model_prefix = 17;       // Prefix for the best model
    string early_stopping_best_model_output_dir = 21;   // Where to store the best model
}

message DoubleMatrix {
    uint32 rows = 1;
    uint32 cols = 2;
    repeated double data = 3 [packed=true];
}

message IntVec2D {
    int32 x = 1;
    int32 y = 2;
}


message CTCDecoderParams {
    enum CTCDecoderType {
        CTC_DEFAULT = 0;
        CTC_TOKEN_PASSING = 1;
        CTC_WORD_BEAM_SEARCH = 2;
    }

    CTCDecoderType type = 1;

    int32 blank_index = 2;
    float min_p_threshold = 3;

    int32 beam_width = 4;
    repeated string non_word_chars = 8;

    repeated string dictionary = 6;

    string word_separator = 7;
}

message NetworkParams {
    enum SolverType {
        MOMENTUM_SOLVER = 0;
        ADAM_SOLVER = 1;
    }

    BackendParams backend = 8;

    int32 channels = 14;                    // e.g. 1: binary/gray, 3: color, or any other number (default 0 equiv 1)
    int32 features = 6;

}

// Backend data
message BackendParams {
    int32 shuffle_buffer_size = 3;      // upper limit for the shuffle buffer used during training, 0 means no limit
    int32 num_inter_threads = 4;
    int32 num_intra_threads = 5;
}

// Data required to run a model
message ModelParams {
    NetworkParams network = 2;
    int32 line_height = 3;
    DataPreprocessorParams data_preprocessor = 4;
    TextProcessorParams text_preprocessor = 5;
    TextProcessorParams text_postprocessor = 6;
    CodecParams codec = 7;
}

message DataPreprocessorParams {
    enum Type {
        DEFAULT_NORMALIZER = 0;
        MULTI_NORMALIZER = 1;
        NOOP_NORMALIZER = 2;
        RANGE_NORMALIZER = 3;
        CENTER_NORMALIZER = 4;
        FINAL_PREPARATION = 5;
        SCALE_TO_HEIGHT = 6;
    }

    enum Binarization {
        GRAY = 0;
        OTSU = 1;
    }

    Type type = 1;
    repeated DataPreprocessorParams children = 2;

    int32 line_height = 3;
    int32 range = 4;
    int32 smoothness = 5;
    int32 pad = 6;
    Binarization binarization = 7;

    bool no_normalize = 8;
    bool no_invert = 9;
    bool no_transpose = 10;
    int32 pad_value = 11;
}

message TextNormalizerReplacementParams {
    string old = 1;
    string new = 2;
    bool regex = 3;     // Match as reqular expression
}

message TextProcessorParams {
    enum Type {
        DEFAULT_NORMALIZER = 0;
        DEFAULT_PRE_NORMALIZER = 1;
        DEFAULT_POST_NORMALIZER = 2;
        MULTI_NORMALIZER = 3;
        NOOP_NORMALIZER = 4;
        STRIP_NORMALIZER = 5;
        BIDI_NORMALIZER = 6;
        TEXT_NORMALIZER = 7;
        TEXT_REGULARIZER = 8;
        STR_TO_CHAR_LIST = 9;
    }

    Type type = 1;
    repeated TextProcessorParams children = 2;

    // Bidirectional text processor
    enum BidiDirection {
        BIDI_AUTO = 0;
        BIDI_LTR = 1;
        BIDI_RTL = 2;
    }
    BidiDirection bidi_direction = 3;

    // normalizer (see https://docs.python.org/3/library/unicodedata.html)
    enum UnicodeNormalizationType {
        NFC = 0;
        NFKC = 1;
        NFD = 2;
        NFKD = 3;
    }
    UnicodeNormalizationType unicode_normalization = 4;
    repeated TextNormalizerReplacementParams replacements = 5;

    // str to char available characters
    repeated string characters = 6;
}


message CodecParams {
    repeated string charset = 1;
}


message VoterParams {
    enum Type {
        SEQUENCE_VOTER = 0;
        CONFIDENCE_VOTER_DEFAULT_CTC = 1;
        CONFIDENCE_VOTER_FUZZY_CTC = 2;
    }

    Type type = 1;
    int32 blank_index = 2;
}

message PredictionCharacter {
    string char = 1;
    int32 label = 2;
    float probability = 3;
}

message PredictionPosition {
    repeated PredictionCharacter chars = 1;         // All characters at this position with its individual probabilities
    int32 local_start = 2;                          // Start of the character in the prediction
    int32 local_end = 3;                            // End of the character in the prediction
    int32 global_start = 4;                         // Start of the character in the image (this is usually the center of a letter)
    int32 global_end = 5;                           // End of the character in the image (this is usually the center of a letter)
}

message Prediction {
    string id = 1;                                  // Use this as hint when saving several predictions into one file
    string sentence = 2;                            // Postprocessed sentence
    repeated int32 labels = 3;                      // [Optional] The labels (depending on the codec)
    repeated PredictionPosition positions = 4;      // The individual predictions for each position
    DoubleMatrix logits = 5;                        // The full prediction matrix
    float total_probability = 6;
    float avg_char_probability = 9;                 // Average character probability
    bool is_voted_result = 7;                       // A voted result does not have logits

    string line_path = 8;                           // [Optional] The original path to the corresponding image file
}

message Predictions {
    repeated Prediction predictions = 1;
    string line_path = 2;
}

message LineGeneratorParameters {
    repeated string fonts = 1;                             // which fonts to use
    int32 font_size = 2;
    float min_script_offset = 3;
    float max_script_offset = 4;
}

message TextGeneratorParameters {
    float word_length_mean = 1;
    float word_length_sigma = 2;

    repeated string charset = 3;
    repeated string super_charset = 14;
    repeated string sub_charset = 15;

    float number_of_words_mean = 4;
    float number_of_words_sigma = 5;
    string word_separator = 6;
    float sub_script_p = 7;
    float super_script_p = 8;
    float bold_p = 9;
    float italic_p = 10;

    float letter_spacing_p = 11;
    float letter_spacing_mean = 12;
    float letter_spacing_sigma = 13;
}
